### PR TITLE
[MXNet-1375][Fit API]Added RNN integration test for fit() API

### PR DESCRIPTION
**Pull request by karan6181**
## Description ##
This PR adds nightly RNN Integration test using Gluon fit() API.
JIRA epic can be found here: https://issues.apache.org/jira/browse/MXNET-1375

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-1375](https://issues.apache.org/jira/browse/MXNET-1375), 
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] test_sentiment_rnn: Test the RNN model on CPU using Synthetic data and on GPU using IMDB dataset. 

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

@roywei @nswamy @abhinavs95

### CODE REVIEW

File: tests/nightly/JenkinsfileForBinaries [hunk 1]
```diff
@@ -106,6 +106,22 @@ core_logic: {
           utils.docker_run('ubuntu_nightly_gpu', 'nightly_tutorial_test_ubuntu_python3_gpu', true, '1500m')
         }
       }
+    },
```
Comment (piyushghai) [inline tests/nightly/JenkinsfileForBinaries]: This should not be in the JenkinsfileForBinaries.

File: tests/nightly/estimator/test_sentiment_rnn.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        encoding = nd.concat(states[0], states[-1])
+        outputs = self.decoder(encoding)
+        return outputs
+
+
+def download_imdb(data_dir='./data'):
+    '''
+    Download and extract the IMDB dataset
+    '''
+    url = ('http://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz')
```
Comment (piyushghai) [inline tests/nightly/estimator/test_sentiment_rnn.py]: Should we keep this data on an S3 bucket instead for more reliability of it being downloaded ?

File: tests/nightly/estimator/test_sentiment_rnn.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        states = self.encoder(embeddings)
+        # Concatenate the hidden states of the initial time step and final
+        # time step to use as the input of the fully connected layer. Its
+        # shape is (batch size, 4 * number of hidden units)
+        encoding = nd.concat(states[0], states[-1])
+        outputs = self.decoder(encoding)
+        return outputs
+
+
+def download_imdb(data_dir='./data'):
```
Comment (piyushghai) [inline tests/nightly/estimator/test_sentiment_rnn.py]: Can you keep this path as /tmp/data ?

File: tests/nightly/estimator/test_sentiment_rnn.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    Test estimator by training Bidirectional RNN for 5 epochs on the IMDB dataset
+    and verify accuracy
+    '''
+    batch_size = 64
+    num_epochs = 5
+    lr = 0.01
+    embed_size = 100
+
+    # Set context
+    if mx.context.num_gpus() > 0:
```
Comment (piyushghai) [inline tests/nightly/estimator/test_sentiment_rnn.py]: Can write this more succinctly as ```ctx = mx.gpu() if mx.context.num_gpus() > 0 else mx.cpu()```

File: tests/nightly/estimator/test_sentiment_rnn.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+        # Define estimator
+        e = est.Estimator(net=net, loss=loss, metrics=acc,
+                          trainers=trainer, context=context)
+        # Begin training
+        e.fit(train_data=train_dataloader, val_data=val_dataloader,
+              epochs=num_epochs, batch_size=batch_size)
+
+
+def test_estimator_gpu():
+    '''
```
Comment (piyushghai) [inline tests/nightly/estimator/test_sentiment_rnn.py]: There are a lot of things in common between ```test_estimator_cpu``` and ```test_estimator_gpu``` methods. 
Both these methods can be clubbed into a single one, with just the context being different. 

I see in ```test_estimator_gpu``` you are using the assertion as well, so maybe your (new) method : ```test_esimator``` (which is same for cpu/gpu) can return an estimator, and you can do your assertions only if being run on gpu context.

**Comment by abhinavs95** [comment]
@mxnet-label-bot add [Gluon, Test]

File: tests/nightly/estimator/test_sentiment_rnn.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+import argparse
+import os
+import tarfile
+import random
+import collections
+import mxnet as mx
+from mxnet import nd
+from mxnet.contrib import text
+from mxnet.gluon import data as gdata, loss as gloss, utils as gutils, nn, rnn
```
Comment (roywei) [inline tests/nightly/estimator/test_sentiment_rnn.py]: just import data, loss, utils, don't rename it. the book did it for separation of mxnet and gluoncv/nlp, d2l package, we only have mxnet here

File: tests/nightly/estimator/test_sentiment_rnn.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    e = est.Estimator(net=net, loss=loss, metrics=acc,
+                      trainers=trainer, context=ctx)
+    # Begin training
+    e.fit(train_data=train_dataloader, val_data=test_dataloader,
+          epochs=num_epochs)
+
+    assert e.train_stats['train_' + acc.name][num_epochs - 1] > 0.70
+
+
+if __name__ == '__main__':
```
Comment (roywei) [inline tests/nightly/estimator/test_sentiment_rnn.py]: you don't need main in integration tests

File: ci/docker/runtime_functions.sh [hunk 8]
```diff
@@ -1296,6 +1296,20 @@ nightly_scala_demo_test_cpu() {
     bash bin/run_im.sh
 }
 
+nightly_estimator_rnn_gpu() {
+    set -ex
+    cd /work/mxnet/tests/nightly/estimator
+    export PYTHONPATH=/work/mxnet/python/
+    python test_sentiment_rnn.py --type gpu
```
Comment (roywei) [inline ci/docker/runtime_functions.sh]: use nosetest and assert accuracy for gpu, do not run python scripts here

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 2 above)
Comment (abhinavs95) [reply tests/nightly/estimator/test_sentiment_rnn.py]: +1 or maybe look if its there in gluon-nlp

File: tests/nightly/estimator/test_sentiment_rnn.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+
+    train_dataloader = gdata.DataLoader(dataset=gdata.ArrayDataset(train_data, train_label),
+                                        batch_size=batch_size, shuffle=True)
+    val_dataloader = gdata.DataLoader(dataset=gdata.ArrayDataset(val_data, val_label),
+                                      batch_size=batch_size)
+    vocab_list = mx.nd.zeros(shape=(100,))
+
+    # Get the model
+    for model in models:
+        if model == 'TextCNN':
```
Comment (abhinavs95) [inline tests/nightly/estimator/test_sentiment_rnn.py]: To avoid splitting logic here we can have a get_model() function which returns the net and related items, and use it for both the tests

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 3 above)
Comment (abhinavs95) [inline tests/nightly/estimator/test_sentiment_rnn.py]: There are 5 functions for loading and processing the dataset, can't they be combined into one?

File: tests/nightly/JenkinsfileForBinaries (hunk 1 above)
Comment (karan6181) [reply tests/nightly/JenkinsfileForBinaries]: Done. Moved to `tests/nightly/Jenkinsfile`. Thank you for your input.

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 3 above)
Comment (karan6181) [reply tests/nightly/estimator/test_sentiment_rnn.py]: Done. Commit [ae81b08](https://github.com/apache/incubator-mxnet/pull/14547/commits/ae81b08fd94686d0501fe4e3c3762f3fef154db5)

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 4 above)
Comment (karan6181) [reply tests/nightly/estimator/test_sentiment_rnn.py]: Yes absolutely. Modified it. [ae81b08](https://github.com/apache/incubator-mxnet/pull/14547/commits/ae81b08fd94686d0501fe4e3c3762f3fef154db5)

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 6 above)
Comment (karan6181) [reply tests/nightly/estimator/test_sentiment_rnn.py]: Thanks. Updated it. [ae81b08](https://github.com/apache/incubator-mxnet/pull/14547/commits/ae81b08fd94686d0501fe4e3c3762f3fef154db5)

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 7 above)
Comment (karan6181) [reply tests/nightly/estimator/test_sentiment_rnn.py]: Updated. [ae81b08](https://github.com/apache/incubator-mxnet/pull/14547/commits/ae81b08fd94686d0501fe4e3c3762f3fef154db5)

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 2 above)
Comment (karan6181) [reply tests/nightly/estimator/test_sentiment_rnn.py]: We have to handle the licensing part if we add it to our S3 bucket. Let's keep it this way for now. I also found this R example  `example/rnn/bucket_R/data_preprocessing_seq_to_one.R` which also uses the data from the same source.

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 5 above)
Comment (karan6181) [reply tests/nightly/estimator/test_sentiment_rnn.py]: Modified a code to reduce the redundancy. Could you please review it again? Thanks!

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 3 above)
Comment (karan6181) [reply tests/nightly/estimator/test_sentiment_rnn.py]: Keeping it separate makes the code more readable and easy to navigate the flow.

File: tests/nightly/estimator/test_sentiment_rnn.py (hunk 9 above)
Comment (karan6181) [reply tests/nightly/estimator/test_sentiment_rnn.py]: We have a separate model class for both the model and we are instantiating the object based on the selection. The `cpu` test runs for both models, however, `gpu` test runs only `BiRNN`.
